File: transpiler/javatests/com/google/j2cl/integration/java/casttoprimitives/Main.java
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package casttoprimitives;

import static com.google.j2cl.integration.testing.Asserts.assertThrowsClassCastException;
import static com.google.j2cl.integration.testing.Asserts.assertTrue;

@SuppressWarnings({"BoxedPrimitiveConstructor", "unchecked"})
public class Main {
  public static void main(String... args) {
    testPrimitiveToPrimitive();
    testObjectReferenceToPrimitive();
    testBoxedToPrimitive();
    testTypeExtendsLongReferenceToPrimitive();
    testTypeExtendsIntersectionReferenceToPrimitive();
    testPrimitiveToReference();
  }

  @SuppressWarnings("unused")
  public static void testPrimitiveToPrimitive() {
    byte b = 1;
    byte maxByte = 127; // Byte.MAX_VALUE
    char c = 'a';
    char maxChar = 65535; // Character.MAX_VALUE;
    short s = 2;
    short maxShort = 32767; // Short.MAX_VALUE;
    int i = 3;
    int maxInt = 2147483647; // Integer.MAX_VALUE;
    long l = 4L;
    long ll = 2415919103L; // max_int < ll < max_int * 2, used for testing for signs.
    long maxLong = 9223372036854775807L; // Long.MAX_VALUE;
    float f = 2.7f;
    float maxFloat = 3.4028235E38f; // Float.MAX_VALUE;
    float posInfiniteFloat = Float.POSITIVE_INFINITY;
    float negInfiniteFloat = Float.NEGATIVE_INFINITY;
    float nanFloat = Float.NaN;
    double d = 2.6;
    double dd = 2415919103.7; // dd > max_int
    double maxDouble = 1.7976931348623157E308; // Double.MAX_VALUE;
    double posInfiniteDouble = Double.POSITIVE_INFINITY;
    double negInfiniteDouble = Double.NEGATIVE_INFINITY;
    double nanDouble = Double.NaN;

    assertTrue(((char) b == 1));
    assertTrue(((short) b == 1));
    assertTrue(((int) b == 1));
    assertTrue(((long) b == 1L));
    assertTrue(((float) b == 1.0));
    assertTrue(((double) b == 1.0));

    assertTrue(((char) maxByte == 127));
    assertTrue(((short) maxByte == 127));
    assertTrue(((int) maxByte == 127));
    assertTrue(((long) maxByte == 127L));
    assertTrue(((float) maxByte == 127.0));
    assertTrue(((double) maxByte == 127.0));

    assertTrue(((byte) c == 97));
    assertTrue(((short) c == 97));
    assertTrue(((int) c == 97));
    assertTrue(((long) c == 97L));
    assertTrue(((float) c == 97.0));
    assertTrue(((double) c == 97.0));

    assertTrue(((byte) maxChar == -1));
    assertTrue(((short) maxChar == -1));
    assertTrue(((int) maxChar == 65535));
    assertTrue(((long) maxChar == 65535L));
    assertTrue(((float) maxChar == 65535.0));
    assertTrue(((double) maxChar == 65535.0));

    assertTrue(((byte) s == 2));
    assertTrue(((char) s == 2));
    assertTrue(((int) s == 2));
    assertTrue(((long) s == 2L));
    assertTrue(((float) s == 2.0));
    assertTrue(((double) s == 2.0));

    assertTrue(((byte) maxShort == -1));
    assertTrue(((char) maxShort == 32767));
    assertTrue(((int) maxShort == 32767));
    assertTrue(((long) maxShort == 32767L));
    assertTrue(((float) maxShort == 32767.0));
    assertTrue(((double) maxShort == 32767.0));

    assertTrue(((byte) i == 3));
    assertTrue(((char) i == 3));
    assertTrue(((short) i == 3));
    assertTrue(((long) i == 3L));
    assertTrue(((float) i == 3.0));
    assertTrue(((double) i == 3.0));

    assertTrue(((byte) maxInt == -1));
    assertTrue(((char) maxInt == 65535));
    assertTrue(((short) maxInt == -1));
    assertTrue(((long) maxInt == 2147483647L));
    // assertTrue(((float) maxInt == 2.14748365E9f)); // it is casted to double-precision
    // 2.147483647E9
    assertTrue(((double) maxInt == 2.147483647E9));

    assertTrue(((byte) l == 4));
    assertTrue(((char) l == 4));
    assertTrue(((short) l == 4));
    assertTrue(((int) l == 4));
    assertTrue(((float) l == 4));
    assertTrue(((double) l == 4));

    assertTrue(((byte) ll == -1));
    assertTrue(((char) ll == 65535));
    assertTrue(((short) ll == -1));
    assertTrue(((int) ll == -1879048193));
    // assertTrue(((float) ll == 2.4159191E9)); // it is casted to double-precision 2.415919103E9.
    assertTrue(((double) ll == 2.415919103E9));

    assertTrue(((byte) maxLong == -1));
    assertTrue(((char) maxLong == 65535));
    assertTrue(((short) maxLong == -1));
    assertTrue(((int) maxLong == -1));
    assertTrue(((float) maxLong == 9.223372036854776E18));
    assertTrue(((double) maxLong == 9.223372036854776E18));

    assertTrue(((byte) f == 2));
    assertTrue(((char) f == 2));
    assertTrue(((short) f == 2));
    assertTrue(((int) f == 2));
    assertTrue(((long) f == 2L));
    assertTrue((((double) f - 2.7) < 1e-7));

    assertTrue(((byte) maxFloat == -1));
    assertTrue(((char) maxFloat == 65535));
    assertTrue(((short) maxFloat == -1));
    assertTrue(((int) maxFloat == 2147483647));
    assertTrue(((long) maxFloat == 9223372036854775807L));
    assertTrue(((double) maxFloat == 3.4028234663852886E38));

    assertTrue(((byte) posInfiniteFloat == -1));
    assertTrue(((short) posInfiniteFloat == -1));
    assertTrue(((char) posInfiniteFloat == 65535));
    assertTrue(((int) posInfiniteFloat == 2147483647));
    assertTrue(((long) posInfiniteFloat == 9223372036854775807L));
    assertTrue(((double) posInfiniteFloat == Double.POSITIVE_INFINITY));

    assertTrue(((byte) negInfiniteFloat == 0));
    assertTrue(((short) negInfiniteFloat == 0));
    assertTrue(((char) negInfiniteFloat == 0));
    assertTrue(((int) negInfiniteFloat == -2147483648));
    assertTrue(((long) negInfiniteFloat == -9223372036854775808L));
    assertTrue(((double) negInfiniteFloat == Double.NEGATIVE_INFINITY));

    assertTrue(((byte) nanFloat == 0));
    assertTrue(((short) nanFloat == 0));
    assertTrue(((char) nanFloat == 0));
    assertTrue(((int) nanFloat == 0));
    assertTrue(((long) nanFloat == 0L));
    assertTrue((Double.isNaN((double) nanFloat)));

    assertTrue(((byte) d == 2));
    assertTrue(((char) d == 2));
    assertTrue(((short) d == 2));
    assertTrue(((int) d == 2));
    assertTrue(((long) d == 2L));
    // assertTrue(((float) d == 2.6f)); // float is emitted as 2.5999999046325684

    assertTrue(((byte) dd == -1));
    assertTrue(((char) dd == 65535));
    assertTrue(((short) dd == -1));
    assertTrue(((int) dd == 2147483647));
    assertTrue(((long) dd == 2415919103L));
    // assertTrue(((float) dd == 2.4159191E9)); // it is casted to double-precision 2415919103.7

    assertTrue(((byte) maxDouble == -1));
    assertTrue(((char) maxDouble == 65535));
    assertTrue(((short) maxDouble == -1));
    assertTrue(((int) maxDouble == 2147483647));
    assertTrue(((long) maxDouble == 9223372036854775807L));

    assertTrue(((byte) posInfiniteDouble == -1));
    assertTrue(((short) posInfiniteDouble == -1));
    assertTrue(((char) posInfiniteDouble == 65535));
    assertTrue(((int) posInfiniteDouble == 2147483647));
    assertTrue(((long) posInfiniteDouble == 9223372036854775807L));
    assertTrue(((float) posInfiniteDouble == Float.POSITIVE_INFINITY));

    assertTrue(((byte) negInfiniteDouble == 0));
    assertTrue(((short) negInfiniteDouble == 0));
    assertTrue(((char) negInfiniteDouble == 0));
    assertTrue(((int) negInfiniteDouble == -2147483648));
    assertTrue(((long) negInfiniteDouble == -9223372036854775808L));
    assertTrue(((float) negInfiniteDouble == Float.NEGATIVE_INFINITY));

    assertTrue(((byte) nanDouble == 0));
    assertTrue(((short) nanDouble == 0));
    assertTrue(((char) nanDouble == 0));
    assertTrue(((int) nanDouble == 0));
    assertTrue(((long) nanDouble == 0L));
    assertTrue((Float.isNaN((float) nanDouble)));
  }

  @SuppressWarnings("unused")
  public static void testObjectReferenceToPrimitive() {
    {
    Object o = new Object();
      assertThrowsClassCastException(
          () -> {
            boolean b = (boolean) o;
          });
      assertThrowsClassCastException(
          () -> {
            byte b = (byte) o;
          });
      assertThrowsClassCastException(
          () -> {
            char c = (char) o;
          });
      assertThrowsClassCastException(
          () -> {
            short s = (short) o;
          });
      assertThrowsClassCastException(
          () -> {
            int i = (int) o;
          });
      assertThrowsClassCastException(
          () -> {
            long l = (long) o;
          });
      assertThrowsClassCastException(
          () -> {
            float f = (float) o;
          });
      assertThrowsClassCastException(
          () -> {
            double d = (double) o;
          });
    }
    {
      Object o = Boolean.FALSE;
      boolean bool = (boolean) o;
      assertTrue(!bool);

      o = Byte.MAX_VALUE;
      byte b = (byte) o;
      assertTrue(b == 127);

      o = new Character('a');
      char c = (char) o;
      assertTrue(c == 'a');

      o = Short.MAX_VALUE;
      short s = (short) o;
      assertTrue(s == 32767);

      o = new Integer(1);
      int i = (int) o;
      assertTrue(i == 1);

      o = new Long(1L);
      long l = (long) o;
      assertTrue(l == 1L);

      o = new Float(1.1f);
      float f = (float) o;
      assertTrue(f == 1.1f);

      o = new Double(1.2);
      double d = (double) o;
      assertTrue(d == 1.2);
    }
  }

  public static <T extends Long> void testTypeExtendsLongReferenceToPrimitive() {
    T o = (T) new Long(1);
    long l = (long) o;
    assertTrue(l == 1);

    float f = (float) o;
    assertTrue(f == 1.0);

    double d = (double) o;
    assertTrue(d == 1.0);
  }

  public static <T extends Long & Comparable<Long>>
      void testTypeExtendsIntersectionReferenceToPrimitive() {
    T o = (T) new Long(1);
    long l = (long) o;
    assertTrue(l == 1);

    float f = (float) o;
    assertTrue(f == 1.0);

    double d = (double) o;
    assertTrue(d == 1.0);
  }

  @SuppressWarnings("unused")
  public static void testBoxedToPrimitive() {
    Byte b = new Byte((byte) 1);
    Character c = new Character('a');
    Short s = new Short((short) 1);
    Integer i = new Integer(1);
    Long l = new Long(1L);
    Float f = new Float(1.1f);
    Double d = new Double(1.1);

    short ss = b;
    assertTrue(ss == 1);

    int ii = b;
    assertTrue(ii == 1);
    ii = c;
    assertTrue(ii == 97);
    ii = s;
    assertTrue(ii == 1);

    long ll = b;
    assertTrue(ll == 1L);
    ll = s;
    assertTrue(ll == 1L);
    ll = c;
    assertTrue(ll == 97L);
    ll = i;
    assertTrue(ll == 1L);

    float ff = b;
    assertTrue(ff == 1);
    ff = s;
    assertTrue(ff == 1);
    ff = c;
    assertTrue(ff == 97);
    ff = i;
    assertTrue(ff == 1);
    ff = l;
    assertTrue(ff == 1);

    double dd = b;
    assertTrue(dd == 1);
    dd = s;
    assertTrue(dd == 1);
    dd = c;
    assertTrue(dd == 97);
    dd = i;
    assertTrue(dd == 1);
    dd = l;
    assertTrue(dd == 1);
    dd = f;
    assertTrue(dd - 1.1 < 1e-7);
  }

  @SuppressWarnings("cast")
  public static void testPrimitiveToReference() {
    boolean bool = true;
    byte b = 1;
    char c = 'a';
    short s = 1;
    int i = 1;
    long l = 1L;
    float f = 1.0f;
    double d = 1.0;
    Object o = bool;
    assertTrue(o != null);
    o = b;
    assertTrue(o != null);
    o = c;
    assertTrue(o != null);
    o = s;
    assertTrue(o != null);
    o = i;
    assertTrue(o != null);
    o = l;
    assertTrue(o != null);
    o = f;
    assertTrue(o != null);
    o = d;
    assertTrue(o != null);
    o = (Object) bool;
    assertTrue(o != null);
    o = (Object) b;
    assertTrue(o != null);
    o = (Object) c;
    assertTrue(o != null);
    o = (Object) s;
    assertTrue(o != null);
    o = (Object) i;
    assertTrue(o != null);
    o = (Object) l;
    assertTrue(o != null);
    o = (Object) f;
    assertTrue(o != null);
    o = (Object) d;
    assertTrue(o != null);
  }
}

File: transpiler/javatests/com/google/j2cl/integration/kotlin/casttoprimitives/Main.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package casttoprimitives

import com.google.j2cl.integration.testing.Asserts.assertThrowsClassCastException
import com.google.j2cl.integration.testing.Asserts.assertTrue

fun main(vararg unused: String) {
  testPrimitiveToPrimitive()
  testObjectReferenceToPrimitive()
  testBoxedToPrimitive()
  testTypeExtendsLongReferenceToPrimitive<Long>()
  testTypeExtendsIntersectionReferenceToPrimitive<Long>()
  testPrimitiveToReference()
}

// Constants that can not be expressed as literals.
private val BYTE_MINUS_ONE: Byte = -1
private val BYTE_ZERO: Byte = 0
private val BYTE_ONE: Byte = 1
private val BYTE_TWO: Byte = 2
private val BYTE_THREE: Byte = 3
private val BYTE_FOUR: Byte = 4
private val BYTE_97: Byte = 97
private val SHORT_MINUS_ONE: Short = -1
private val SHORT_ZERO: Short = 0
private val SHORT_ONE: Short = 1
private val SHORT_TWO: Short = 2
private val SHORT_THREE: Short = 3
private val SHORT_FOUR: Short = 4
private val SHORT_97: Short = 97
private val SHORT_127: Short = 127

fun testPrimitiveToPrimitive() {
  val b: Byte = 1
  val maxByte: Byte = 127 // Byte.MAX_VALUE
  val c: Char = 'a'
  val maxChar: Char = '\uFFFF' // Character.MAX_VALUE;
  val s: Short = 2
  val maxShort: Short = 32767 // Short.MAX_VALUE
  val i: Int = 3
  val maxInt: Int = 2147483647 // Integer.MAX_VALUE
  val l: Long = 4L
  val ll: Long = 2415919103L // max_int < ll < max_int * 2, used for testing for signs.
  val maxLong: Long = 9223372036854775807L // Long.MAX_VALUE;
  val f: Float = 2.7f
  val maxFloat: Float = 3.4028235E38f // Float.MAX_VALUE;
  val posInfiniteFloat: Float = Float.POSITIVE_INFINITY
  val negInfiniteFloat: Float = Float.NEGATIVE_INFINITY
  val nanFloat: Float = Float.NaN
  val d: Double = 2.6
  val dd: Double = 2415919103.7 // dd > max_int
  val maxDouble: Double = 1.7976931348623157E308 // Double.MAX_VALUE;
  val posInfiniteDouble: Double = Double.POSITIVE_INFINITY
  val negInfiniteDouble: Double = Double.NEGATIVE_INFINITY
  val nanDouble: Double = Double.NaN

  assertTrue(b.toChar() == '\u0001')
  assertTrue(b.toShort() == SHORT_ONE)
  assertTrue(b.toInt() == 1)
  assertTrue(b.toLong() == 1L)
  assertTrue(b.toFloat() == 1.0f)
  assertTrue(b.toDouble() == 1.0)

  assertTrue(maxByte.toChar() == '\u007F')
  assertTrue(maxByte.toShort() == SHORT_127)
  assertTrue(maxByte.toInt() == 127)
  assertTrue(maxByte.toLong() == 127L)
  assertTrue(maxByte.toFloat() == 127.0f)
  assertTrue(maxByte.toDouble() == 127.0)

  assertTrue(c.toByte() == BYTE_97)
  assertTrue(c.toShort() == SHORT_97)
  assertTrue(c == '\u0061')
  assertTrue(c.toLong() == 97L)
  assertTrue(c.toFloat() == 97.0f)
  assertTrue(c.toDouble() == 97.0)
  assertTrue(c.toChar() == 'a')

  assertTrue(maxChar.toByte() == BYTE_MINUS_ONE)
  assertTrue(maxChar.toShort() == SHORT_MINUS_ONE)
  assertTrue(maxChar.toInt() == 65535)
  assertTrue(maxChar.toLong() == 65535L)
  assertTrue(maxChar.toFloat() == 65535.0f)
  assertTrue(maxChar.toDouble() == 65535.0)

  assertTrue(s.toByte() == BYTE_TWO)
  assertTrue(s.toChar() == '\u0002')
  assertTrue(s.toInt() == 2)
  assertTrue(s.toLong() == 2L)
  assertTrue(s.toFloat() == 2.0f)
  assertTrue(s.toDouble() == 2.0)

  assertTrue(maxShort.toByte() == BYTE_MINUS_ONE)
  assertTrue(maxShort.toChar() == '\u7FFF')
  assertTrue(maxShort.toInt() == 32767)
  assertTrue(maxShort.toLong() == 32767L)
  assertTrue(maxShort.toFloat() == 32767.0f)
  assertTrue(maxShort.toDouble() == 32767.0)

  assertTrue(i.toByte() == BYTE_THREE)
  assertTrue(i.toChar() == '\u0003')
  assertTrue(i.toShort() == SHORT_THREE)
  assertTrue(i.toLong() == 3L)
  assertTrue(i.toFloat() == 3.0f)
  assertTrue(i.toDouble() == 3.0)

  assertTrue(maxInt.toByte() == BYTE_MINUS_ONE)
  assertTrue(maxInt.toChar() == '\uFFFF')
  assertTrue(maxInt.toShort() == SHORT_MINUS_ONE)
  assertTrue(maxInt.toLong() == 2147483647L)
  // assertTrue(maxInt.toFloat() == 2.14748365E9f)); // toFloat() returns double precision
  // 2.147483647E9
  assertTrue(maxInt.toDouble() == 2.147483647E9)

  assertTrue(l.toByte() == BYTE_FOUR)
  assertTrue(l.toChar() == '\u0004')
  assertTrue(l.toShort() == SHORT_FOUR)
  assertTrue(l.toInt() == 4)
  assertTrue(l.toFloat() == 4f)
  assertTrue(l.toDouble() == 4.0)

  assertTrue(ll.toByte() == BYTE_MINUS_ONE)
  assertTrue(ll.toChar() == '\uFFFF')
  assertTrue(ll.toShort() == SHORT_MINUS_ONE)
  assertTrue(ll.toInt() == -1879048193)
  // assertTrue(ll.toFloat() == 2.4159191E9)); // toFloat() returns double-precision 2.415919103E9.
  assertTrue(ll.toDouble() == 2.415919103E9)

  assertTrue(maxLong.toByte() == BYTE_MINUS_ONE)
  assertTrue(maxLong.toChar() == '\uFFFF')
  assertTrue(maxLong.toShort() == SHORT_MINUS_ONE)
  assertTrue(maxLong.toInt() == -1)
  assertTrue(maxLong.toFloat() == 9.223372036854776E18f)
  assertTrue(maxLong.toDouble() == 9.223372036854776E18)

  assertTrue(f.toInt().toByte() == BYTE_TWO)
  assertTrue(f.toInt().toChar() == '\u0002')
  assertTrue(f.toInt().toShort() == SHORT_TWO)
  assertTrue(f.toInt() == 2)
  assertTrue(f.toLong() == 2L)
  assertTrue(f.toDouble() - 2.7 < 1e-7)

  assertTrue(maxFloat.toInt().toByte() == BYTE_MINUS_ONE)
  assertTrue(maxFloat.toChar() == '\uFFFF')
  assertTrue(maxFloat.toInt().toShort() == SHORT_MINUS_ONE)
  assertTrue(maxFloat.toInt() == 2147483647)
  assertTrue(maxFloat.toLong() == 9223372036854775807L)
  assertTrue(maxFloat.toDouble() == 3.4028234663852886E38)

  assertTrue(posInfiniteFloat.toInt().toByte() == BYTE_MINUS_ONE)
  assertTrue(posInfiniteFloat.toInt().toShort() == SHORT_MINUS_ONE)
  assertTrue(posInfiniteFloat.toInt().toChar() == '\uFFFF')
  assertTrue(posInfiniteFloat.toInt() == 2147483647)
  assertTrue(posInfiniteFloat.toLong() == 9223372036854775807L)
  assertTrue(posInfiniteFloat.toDouble() == Double.POSITIVE_INFINITY)

  assertTrue(negInfiniteFloat.toInt().toByte() == BYTE_ZERO)
  assertTrue(negInfiniteFloat.toInt().toShort() == SHORT_ZERO)
  assertTrue(negInfiniteFloat.toInt().toChar() == 0.toChar())
  assertTrue(negInfiniteFloat.toInt() == -2147483648)
  assertTrue(negInfiniteFloat.toLong() == -9223372036854775807L - 1L)
  assertTrue(negInfiniteFloat.toDouble() == Double.NEGATIVE_INFINITY)

  assertTrue(nanFloat.toInt().toByte() == BYTE_ZERO)
  assertTrue(nanFloat.toInt().toShort() == SHORT_ZERO)
  assertTrue(nanFloat.toInt().toChar() == '\u0000')
  assertTrue(nanFloat.toInt() == 0)
  assertTrue(nanFloat.toLong() == 0L)
  assertTrue(java.lang.Double.isNaN(nanFloat.toDouble()))

  assertTrue(d.toInt().toByte() == BYTE_TWO)
  assertTrue(d.toInt().toChar() == '\u0002')
  assertTrue(d.toInt().toShort() == SHORT_TWO)
  assertTrue(d.toInt() == 2)
  assertTrue(d.toLong() == 2L)
  // assertTrue((d.toFloat() == 2.6f)); // float is emitted as 2.5999999046325684

  assertTrue(dd.toInt().toByte() == BYTE_MINUS_ONE)
  assertTrue(dd.toInt().toChar() == '\uFFFF')
  assertTrue(dd.toInt().toShort() == SHORT_MINUS_ONE)
  assertTrue(dd.toInt() == 2147483647)
  assertTrue(dd.toLong() == 2415919103L)
  // assertTrue((dd.toFloat() == 2.4159191E9)); // toFloat() returns double-precision 2415919103.7

  assertTrue(maxDouble.toInt().toByte() == BYTE_MINUS_ONE)
  assertTrue(maxDouble.toInt().toChar() == '\uFFFF')
  assertTrue(maxDouble.toInt().toShort() == SHORT_MINUS_ONE)
  assertTrue(maxDouble.toInt() == 2147483647)
  assertTrue(maxDouble.toLong() == 9223372036854775807L)

  assertTrue(posInfiniteDouble.toInt().toByte() == BYTE_MINUS_ONE)
  assertTrue(posInfiniteDouble.toInt().toShort() == SHORT_MINUS_ONE)
  assertTrue(posInfiniteDouble.toInt().toChar() == '\uFFFF')
  assertTrue(posInfiniteDouble.toInt() == 2147483647)
  assertTrue(posInfiniteDouble.toLong() == 9223372036854775807L)
  assertTrue(posInfiniteDouble.toFloat() == Float.POSITIVE_INFINITY)

  assertTrue(negInfiniteDouble.toInt().toByte() == BYTE_ZERO)
  assertTrue(negInfiniteDouble.toInt().toShort() == SHORT_ZERO)
  assertTrue(negInfiniteDouble.toInt().toChar() == '\u0000')
  assertTrue(negInfiniteDouble.toInt() == -2147483648)
  assertTrue(negInfiniteDouble.toLong() == -9223372036854775807L - 1L)
  assertTrue(negInfiniteDouble.toFloat() == Float.NEGATIVE_INFINITY)

  assertTrue(nanDouble.toInt().toByte() == BYTE_ZERO)
  assertTrue(nanDouble.toInt().toShort() == SHORT_ZERO)
  assertTrue(nanDouble.toInt().toChar() == '\u0000')
  assertTrue(nanDouble.toInt() == 0)
  assertTrue(nanDouble.toLong() == 0L)
  assertTrue(java.lang.Float.isNaN(nanDouble.toFloat()))
}

fun testObjectReferenceToPrimitive() {
  val o: Any? = Any()
  assertThrowsClassCastException {
    val b = o as Boolean
  }
  assertThrowsClassCastException {
    val b = o as Byte
  }
  assertThrowsClassCastException {
    val c = o as Char
  }
  assertThrowsClassCastException {
    val s = o as Short
  }
  assertThrowsClassCastException {
    val i = o as Int
  }
  assertThrowsClassCastException {
    val l = o as Long
  }
  assertThrowsClassCastException {
    val f = o as Float
  }
  assertThrowsClassCastException {
    val d = o as Double
  }

  var o1: Any? = false
  val bool = o1 as Boolean
  assertTrue(!bool)

  o1 = Byte.MAX_VALUE
  assertTrue(o1.toInt() == 127)

  o1 = 'a'
  assertTrue(o1 == 'a')

  o1 = Short.MAX_VALUE
  assertTrue(o1.toInt() == 32767)

  o1 = 1
  assertTrue(o1 == 1)

  o1 = 1L
  assertTrue(o1 == 1L)

  o1 = 1.1f
  assertTrue(o1 == 1.1f)

  o1 = 1.2
  assertTrue(o1 == 1.2)
}

fun <T : Long?> testTypeExtendsLongReferenceToPrimitive() {
  val o = 1L as T
  val l = o as Long
  assertTrue(l == 1L)
  val f = o.toFloat()
  assertTrue(f == 1.0f)
  val d = o.toDouble()
  assertTrue(d == 1.0)
}

fun <T> testTypeExtendsIntersectionReferenceToPrimitive() where T : Long?, T : Comparable<Long>? {
  val o = 1L as T
  val l = o as Long
  assertTrue(l == 1L)
  val f = o.toFloat()
  assertTrue(f == 1.0f)
  val d = o.toDouble()
  assertTrue(d == 1.0)
}

fun testBoxedToPrimitive() {
  val b = BYTE_ONE
  val c = 'a'
  val s = SHORT_ONE
  val i = 1
  val l = 1L
  val f = 1.1f
  val d = 1.1

  val ss = b.toShort()
  assertTrue(ss.toInt() == 1)

  var ii = b.toInt()
  assertTrue(ii == 1)
  ii = c.toInt()
  assertTrue(ii == 97)
  ii = s.toInt()
  assertTrue(ii == 1)

  var ll = b.toLong()
  assertTrue(ll == 1L)
  ll = s.toLong()
  assertTrue(ll == 1L)
  ll = c.toInt().toLong()
  assertTrue(ll == 97L)
  ll = i.toLong()
  assertTrue(ll == 1L)

  var ff = b.toFloat()
  assertTrue(ff == 1f)
  ff = s.toFloat()
  assertTrue(ff == 1f)
  ff = c.toInt().toFloat()
  assertTrue(ff == 97f)
  ff = i.toFloat()
  assertTrue(ff == 1f)
  ff = l.toFloat()
  assertTrue(ff == 1f)

  var dd = b.toDouble()
  assertTrue(dd == 1.0)
  dd = s.toDouble()
  assertTrue(dd == 1.0)
  dd = c.toInt().toDouble()
  assertTrue(dd == 97.0)
  dd = i.toDouble()
  assertTrue(dd == 1.0)
  dd = l.toDouble()
  assertTrue(dd == 1.0)
  dd = f.toDouble()
  assertTrue(dd - 1.1 < 1e-7)
}

fun testPrimitiveToReference() {
  val bool = true
  val b: Byte = 1
  val c = 'a'
  val s: Short = 1
  val i = 1
  val l = 1L
  val f = 1.0f
  val d = 1.0
  var o: Any = bool
  assertTrue(o != null)
  o = b
  assertTrue(o != null)
  o = c
  assertTrue(o != null)
  o = s
  assertTrue(o != null)
  o = i
  assertTrue(o != null)
  o = l
  assertTrue(o != null)
  o = f
  assertTrue(o != null)
  o = d
  assertTrue(o != null)
  o = bool
  assertTrue(o != null)
  o = b
  assertTrue(o != null)
  o = c
  assertTrue(o != null)
  o = s
  assertTrue(o != null)
  o = i
  assertTrue(o != null)
  o = l
  assertTrue(o != null)
  o = f
  assertTrue(o != null)
  o = d
  assertTrue(o != null)
}
